### PR TITLE
Image classification/one shot

### PR DESCRIPTION
**Pull request by rahul-tuli**
The goal of this PR is to allow easy one-shot application of recipes to image-classification models.

To accomplish this we add a `--one-shot` argument to `sparseml.image_classification.train` script

The new functionality was tested locally with the following:

```bash
sparseml.image_classification.train \
    --train_batch_size 32 \
    --test_batch_size 32 \
    --dataset imagenette \
    --dataset-path ./datasets/imagenette \
    --arch-key resnet50 \
    --recipe-path zoo:cv/classification/resnet_v1-50/pytorch/sparseml/imagenette/pruned-conservative?recipe_type=original \
    --checkpoint-path zoo \
    --one-shot
```

The Output is as follows:
```bash
2022-05-11 13:19:01 __main__     INFO     running on device cuda
2022-05-11 13:19:01 __main__     INFO     running on device cuda
2022-05-11 13:19:02 sparseml.optim.manager INFO     Created recipe manager with metadata: {
 "__metadata__": null
}
2022-05-11 13:19:02 __main__     INFO     completed...
2022-05-11 13:19:02 __main__     INFO     completed...
Saving model for epoch -1 to /XXXXXXX//src/sparseml/pytorch/image_classification/pytorch_vision/resnet50_imagenette__10 for model-one-shot
2022-05-11 13:19:02 __main__     INFO     layer sparsities:
2022-05-11 13:19:02 __main__     INFO     layer sparsities:
2022-05-11 13:19:02 __main__     INFO     input.conv.weight: 0.0000
2022-05-11 13:19:02 __main__     INFO     input.conv.weight: 0.0000
2022-05-11 13:19:02 __main__     INFO     sections.0.0.conv1.weight: 0.8000
2022-05-11 13:19:02 __main__     INFO     sections.0.0.conv1.weight: 0.8000
2022-05-11 13:19:02 __main__     INFO     sections.0.0.conv2.weight: 0.8000
2022-05-11 13:19:02 __main__     INFO     sections.0.0.conv2.weight: 0.8000
2022-05-11 13:19:02 __main__     INFO     sections.0.0.conv3.weight: 0.8000
2022-05-11 13:19:02 __main__     INFO     sections.0.0.conv3.weight: 0.8000
2022-05-11 13:19:02 __main__     INFO     sections.0.0.identity.conv.weight: 0.8000
2022-05-11 13:19:02 __main__     INFO     sections.0.0.identity.conv.weight: 0.8000
2022-05-11 13:19:02 __main__     INFO     sections.0.1.conv1.weight: 0.8000
2022-05-11 13:19:02 __main__     INFO     sections.0.1.conv1.weight: 0.8000
2022-05-11 13:19:02 __main__     INFO     sections.0.1.conv2.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.0.1.conv2.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.0.1.conv3.weight: 0.8000
2022-05-11 13:19:02 __main__     INFO     sections.0.1.conv3.weight: 0.8000
2022-05-11 13:19:02 __main__     INFO     sections.0.2.conv1.weight: 0.8000
2022-05-11 13:19:02 __main__     INFO     sections.0.2.conv1.weight: 0.8000
2022-05-11 13:19:02 __main__     INFO     sections.0.2.conv2.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.0.2.conv2.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.0.2.conv3.weight: 0.8000
2022-05-11 13:19:02 __main__     INFO     sections.0.2.conv3.weight: 0.8000
2022-05-11 13:19:02 __main__     INFO     sections.1.0.conv1.weight: 0.8000
2022-05-11 13:19:02 __main__     INFO     sections.1.0.conv1.weight: 0.8000
2022-05-11 13:19:02 __main__     INFO     sections.1.0.conv2.weight: 0.9000
2022-05-11 13:19:02 __main__     INFO     sections.1.0.conv2.weight: 0.9000
2022-05-11 13:19:02 __main__     INFO     sections.1.0.conv3.weight: 0.8000
2022-05-11 13:19:02 __main__     INFO     sections.1.0.conv3.weight: 0.8000
2022-05-11 13:19:02 __main__     INFO     sections.1.0.identity.conv.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.1.0.identity.conv.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.1.1.conv1.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.1.1.conv1.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.1.1.conv2.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.1.1.conv2.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.1.1.conv3.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.1.1.conv3.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.1.2.conv1.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.1.2.conv1.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.1.2.conv2.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.1.2.conv2.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.1.2.conv3.weight: 0.8000
2022-05-11 13:19:02 __main__     INFO     sections.1.2.conv3.weight: 0.8000
2022-05-11 13:19:02 __main__     INFO     sections.1.3.conv1.weight: 0.8000
2022-05-11 13:19:02 __main__     INFO     sections.1.3.conv1.weight: 0.8000
2022-05-11 13:19:02 __main__     INFO     sections.1.3.conv2.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.1.3.conv2.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.1.3.conv3.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.1.3.conv3.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.2.0.conv1.weight: 0.8000
2022-05-11 13:19:02 __main__     INFO     sections.2.0.conv1.weight: 0.8000
2022-05-11 13:19:02 __main__     INFO     sections.2.0.conv2.weight: 0.9000
2022-05-11 13:19:02 __main__     INFO     sections.2.0.conv2.weight: 0.9000
2022-05-11 13:19:02 __main__     INFO     sections.2.0.conv3.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.2.0.conv3.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.2.0.identity.conv.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.2.0.identity.conv.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.2.1.conv1.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.2.1.conv1.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.2.1.conv2.weight: 0.9000
2022-05-11 13:19:02 __main__     INFO     sections.2.1.conv2.weight: 0.9000
2022-05-11 13:19:02 __main__     INFO     sections.2.1.conv3.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.2.1.conv3.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.2.2.conv1.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.2.2.conv1.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.2.2.conv2.weight: 0.9000
2022-05-11 13:19:02 __main__     INFO     sections.2.2.conv2.weight: 0.9000
2022-05-11 13:19:02 __main__     INFO     sections.2.2.conv3.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.2.2.conv3.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.2.3.conv1.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.2.3.conv1.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.2.3.conv2.weight: 0.9000
2022-05-11 13:19:02 __main__     INFO     sections.2.3.conv2.weight: 0.9000
2022-05-11 13:19:02 __main__     INFO     sections.2.3.conv3.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.2.3.conv3.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.2.4.conv1.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.2.4.conv1.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.2.4.conv2.weight: 0.9000
2022-05-11 13:19:02 __main__     INFO     sections.2.4.conv2.weight: 0.9000
2022-05-11 13:19:02 __main__     INFO     sections.2.4.conv3.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.2.4.conv3.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.2.5.conv1.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.2.5.conv1.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.2.5.conv2.weight: 0.9000
2022-05-11 13:19:02 __main__     INFO     sections.2.5.conv2.weight: 0.9000
2022-05-11 13:19:02 __main__     INFO     sections.2.5.conv3.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.2.5.conv3.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.3.0.conv1.weight: 0.8000
2022-05-11 13:19:02 __main__     INFO     sections.3.0.conv1.weight: 0.8000
2022-05-11 13:19:02 __main__     INFO     sections.3.0.conv2.weight: 0.9000
2022-05-11 13:19:02 __main__     INFO     sections.3.0.conv2.weight: 0.9000
2022-05-11 13:19:02 __main__     INFO     sections.3.0.conv3.weight: 0.9000
2022-05-11 13:19:02 __main__     INFO     sections.3.0.conv3.weight: 0.9000
2022-05-11 13:19:02 __main__     INFO     sections.3.0.identity.conv.weight: 0.9000
2022-05-11 13:19:02 __main__     INFO     sections.3.0.identity.conv.weight: 0.9000
2022-05-11 13:19:02 __main__     INFO     sections.3.1.conv1.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.3.1.conv1.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.3.1.conv2.weight: 0.9000
2022-05-11 13:19:02 __main__     INFO     sections.3.1.conv2.weight: 0.9000
2022-05-11 13:19:02 __main__     INFO     sections.3.1.conv3.weight: 0.9000
2022-05-11 13:19:02 __main__     INFO     sections.3.1.conv3.weight: 0.9000
2022-05-11 13:19:02 __main__     INFO     sections.3.2.conv1.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.3.2.conv1.weight: 0.8500
2022-05-11 13:19:02 __main__     INFO     sections.3.2.conv2.weight: 0.9000
2022-05-11 13:19:02 __main__     INFO     sections.3.2.conv2.weight: 0.9000
2022-05-11 13:19:02 __main__     INFO     sections.3.2.conv3.weight: 0.9000
2022-05-11 13:19:02 __main__     INFO     sections.3.2.conv3.weight: 0.9000
2022-05-11 13:19:02 __main__     INFO     classifier.fc.weight: 0.8000
2022-05-11 13:19:02 __main__     INFO     classifier.fc.weight: 0.8000

Process finished with exit code 0

```

### CODE REVIEW

File: src/sparseml/pytorch/image_classification/train.py [hunk 1]
```diff
@@ -628,19 +645,20 @@ def train(
...
-    )
+    if not trainer.one_shot:
+        # Baseline eval run
+        trainer.run_one_epoch(
+            mode="validation",
+            max_steps=debug_steps,
+            baseline_run=True,
+        )
 
     if not eval_mode:
```
Comment (bfineran) [inline src/sparseml/pytorch/image_classification/train.py:656]: let's change this to `... and not trainer.one_shot` so all the `starting training...` logging doesn't print for one shot. can move the final save step at the end of this block out with another check for `eval_mode`

File: src/sparseml/pytorch/image_classification/train.py [hunk 2]
```diff
@@ -536,7 +552,7 @@ def main(
         model_kwargs=model_kwargs,
     )
 
-    model, arch_key = helpers.create_model(
+    model, arch_key, checkpoint_path = helpers.create_model(
```
[the remark this answers is not in the document]
Comment (rahul-tuli) [reply src/sparseml/pytorch/image_classification/train.py:555]: It creates the trainer with correct local checkpoint_path, this is included in PR #778 , I'll rebase over that and change, the base branch

**Review by bfineran — CHANGES_REQUESTED** [review]
good catch, left comment on logic flow

File: src/sparseml/pytorch/image_classification/utils/trainer.py [hunk 3]
```diff
@@ -144,6 +147,8 @@ def __init__(
             self.module_trainer = self._initialize_module_trainer()
         else:
             self.optim = self.manager = self.module_trainer = None
+            if self.one_shot:
+                self._one_shot()
```
Comment (bfineran) [inline src/sparseml/pytorch/image_classification/utils/trainer.py:151]: one shot must be applied whether or not there is a `train_loader`

File: src/sparseml/pytorch/image_classification/train.py [hunk 4]
```diff
@@ -557,7 +570,7 @@ def main(
             ffcv=ffcv,
             device=device,
         )
-        if is_main_process
+        if is_main_process and not one_shot
```
Comment (bfineran) [inline src/sparseml/pytorch/image_classification/train.py:573]: should still load datasets if one shot and info provided

File: src/sparseml/pytorch/image_classification/utils/trainer.py [hunk 5]
```diff
@@ -207,6 +213,14 @@ def max_epochs(self):
         """
         return self.manager.max_epochs if self.manager is not None else 0
 
+    def _apply_one_shot(self):
+        self.manager = ScheduledModifierManager.from_yaml(
+            self.recipe_path,
+        )
+
+        self.manager.apply(self.model)
+        _LOGGER.info("Applied recipe to manager")
```
Comment (bfineran) [inline src/sparseml/pytorch/image_classification/utils/trainer.py:222]: nit: change to `applied {recipe} in one-shot`

**Review by bfineran — APPROVED** [review]

**Review by dbogunowicz — APPROVED** [review]

**Pull request merged**